### PR TITLE
Feature/lp 949 update nodejs

### PR DESCRIPTION
**Pull request by alejouribesanchez**
- Refactor SDK to make it granular
- Add W2 Missing functions
- Add W9 functions
- Add W8 functions
- Add Business Card functions
- Add Bank statements functions
- Add Checks functions
- Add Any docs Missing functions
- Add all JS unit tests
- Skip Typescript Unit tests

### CODE REVIEW

File: lib/anydocs/processAnyDocument.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * veryfi_client.process_any_document('file/path','blue_print')
+ *
+ * @memberof Client
+ * @param {String} file_path Path on disk to a file to submit for data extraction
+ * @param {String} template_name name of the extraction templates.
+ * @param {number} max_pages_to_process The number of pages to process for the document. The limit is 50 pages per document.
+ * @param {Object} kwargs Additional request parameters
+ * @returns {JSON} Data extracted from the document
+ */
+Client.prototype.process_any_document = async function (
```
Comment (manycoding) [inline lib/anydocs/processAnyDocument.js:16]: maybe let's just have one function to read file from path and send a multipart?
So it replaces this one and process_any_document_from_buffer
I see no reason to have base64 for example as it's slower

File: lib/businesscards/deleteBusinessCard.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const Client = require('../client/constructor');
+/**
+ * Delete a Business Card from Veryfi. https://docs.veryfi.com/api/business-cards/delete-a-business-card/
+ *
+ * @memberof Client
+ * @param {string} document_id ID of the document you'd like to delete
+ */
+Client.prototype.delete_business_card = async function (document_id) {
+    let endpoint_name = `/business-cards/${document_id}/`;
```
Comment (manycoding) [inline lib/businesscards/deleteBusinessCard.js:9]: is it using v8?

File: lib/client/constructor.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * @param {string} api_version
+ * @param {Number} timeout
+ */
+function Client(
+    client_id,
+    client_secret,
+    username,
+    api_key,
+    base_url = "https://api.veryfi.com/",
+    api_version = "v8",
```
Comment (manycoding) [inline lib/client/constructor.js:18]: let's not allow this to be changed, less chances somebody passes v7

File: lib/client/getHeaders.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+Client.prototype._get_headers = function (has_files = false) {
+    let final_headers = {
+        "User-Agent": "Node.js Veryfi-Nodejs/1.4.0",
+        "Accept": "application/json",
+        "Content-Type": "application/json",
+        "Client-Id": this.client_id,
+        "Authorization": `apikey ${this.username}:${this.api_key}`,
+    };
+    if (has_files) {
+        final_headers["Content-Type"] = "application/x-www-form-urlencoded";
```
Comment (manycoding) [inline lib/client/getHeaders.js:18]: multipart/form-data

File: lib/client/request.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        return await axios.request({
+            method: http_verb,
+            url: api_url,
+            headers: headers,
+            data: request_arguments,
+            timeout: this.timeout * 1000,
+            params: params
+        });
+    } catch (response) {
+        let errorStatus = response?.response?.status;
```
Comment (manycoding) [inline lib/client/request.js:54]: does it confirm to our errors?
We have status, error/message and details.
Message is old one and will be removed at some point.
https://docs.veryfi.com/api/business-cards/process-a-business-card/

File: lib/documents/processDocumentBase64.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * @memberof Client
+ * @param {String} base64_encoded_string Buffer string of a file to submit for data extraction
+ * @param {String} file_name The file name including the extension
+ * @param {Array} categories List of categories Veryfi can use to categorize the document
+ * @param {Boolean} delete_after_processing Delete this document from Veryfi after data has been extracted
+ * @param {Object} kwargs Additional request parameters
+ * @returns {JSON} Data extracted from the document
+ */
+Client.prototype.process_document_base64string = async function (
```
Comment (manycoding) [inline lib/documents/processDocumentBase64.js:19]: I would leave it out, what's the use for a user?

File: lib/documents/processDocumentStream.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * @memberof Client
+ * @param {ReadStream} file ReadStream of a file to submit for data extraction
+ * @param {String} file_name The file name including the extension
+ * @param {Array} categories List of categories Veryfi can use to categorize the document
+ * @param {Boolean} delete_after_processing Delete this document from Veryfi after data has been extracted
+ * @param {Object} kwargs Additional request parameters
+ * @returns {JSON} Data extracted from the document
+ */
+Client.prototype.process_document_stream = async function (
```
Comment (manycoding) [inline lib/documents/processDocumentStream.js:19]: what's the use instead of file_path?

File: lib/documents/processDocumentUrl.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ * @param {Object} kwargs Additional request parameters
+ * @return {JSON} Data extracted from the document.
+ */
+Client.prototype.process_document_url = async function (
+    file_url = "",
+    file_urls = null,
+    categories = null,
+    delete_after_processing = false,
+    boost_mode = 0,
+    external_id = "",
```
Comment (manycoding) [inline lib/documents/processDocumentUrl.js:24]: null

File: lib/types/BillTo.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import {BoundingElement} from "./BoundingElement";
+
+export declare type BillTo = {
```
Comment (manycoding) [inline lib/types/BillTo.ts:3]: How this types work? Do they cover new fields?

File: lib/documents/processDocumentUrl.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ * @param {Array} categories List of categories to use when categorizing the document
+ * @param {boolean} delete_after_processing Delete this document from Veryfi after data has been extracted
+ * @param {int} boost_mode Flag that tells Veryfi whether boost mode should be enabled. When set to 1, Veryfi will skip data enrichment steps, but will process the document faster. Default value for this flag is 0
+ * @param {string} external_id Optional custom document identifier. Use this if you would like to assign your own ID to documents
+ * @param {int} max_pages_to_process When sending a long document to Veryfi for processing, this parameter controls how many pages of the document will be read and processed, starting from page 1.
+ * @param {Object} kwargs Additional request parameters
+ * @return {JSON} Data extracted from the document.
+ */
+Client.prototype.process_document_url = async function (
+    file_url = "",
```
Comment (manycoding) [inline lib/documents/processDocumentUrl.js:19]: null

File: lib/client/constructor.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    timeout = 120) {
+    this.client_id = client_id;
+    this.client_secret = client_secret;
+    this.username = username;
+    this.api_key = api_key;
+    this.base_url = base_url;
+    this.timeout = timeout;
+    this.api_version = api_version;
+    this.CATEGORIES = [
```
Comment (manycoding) [inline lib/client/constructor.js:27]: I would actually remove default categories, don't think they matter anymore

**Review by manycoding — CHANGES_REQUESTED** [review]

File: lib/anydocs/processAnyDocumentBuffer.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ * @param {String} file_name The file name including the extension
+ * @param {String} file_buffer Buffer of a file to submit for data extraction
+ * @param {String} template_name name of the extraction templates.
+ * @param {number} max_pages_to_process The number of pages to process for the document. The limit is 50 pages per document.
+ * @param {Object} kwargs Additional request parameters
+ * @returns {JSON} Data extracted from the document
+ */
+Client.prototype.process_any_document_from_buffer = async function (
+    file_name,
+    file_buffer,
```
Comment (manycoding) [inline lib/anydocs/processAnyDocumentBuffer.js:15]: why file_data and not multipart?

File: lib/businesscards/deleteBusinessCard.js (hunk 2 above)
Comment (alejouribesanchez) [reply lib/businesscards/deleteBusinessCard.js:9]: Yes

File: lib/client/constructor.js (hunk 3 above)
Comment (alejouribesanchez) [reply lib/client/constructor.js:18]: Done

File: lib/client/constructor.js (hunk 11 above)
Comment (alejouribesanchez) [reply lib/client/constructor.js:27]: Done

File: lib/documents/processDocument.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    delete_after_processing = false,
+    {...kwargs} = {}
+) {
+
+    let endpoint_name = "/documents/";
+    if (!categories) {
+        categories = this.CATEGORIES;
+    }
+    let file_name = path.basename(file_path);
+    const image_file = fs.readFileSync(file_path, {encoding: 'base64'});
```
Comment (manycoding) [inline lib/documents/processDocument.js:31]: should be multipart

File: lib/documents/processDocumentBase64.js (hunk 6 above)
Comment (manycoding) [reply lib/documents/processDocumentBase64.js:19]: if you wanna keep it let’s check if data has type hint and reject if not

**Review by manycoding — CHANGES_REQUESTED** [review]

File: lib/anydocs/processAnyDocument.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
```
Comment (manycoding) [inline lib/anydocs/processAnyDocument.js:16]: so why you left this one as base64?

File: lib/anydocs/processAnyDocumentBuffer.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    template_name,
+    max_pages_to_process = 20,
+    {...kwargs} = {}
+) {
+    const file_data = this.add_mime_type(file_buffer, file_name);
+    let endpoint_name = "/any-documents/";
+    let request_arguments = {
+        "file_name": file_name,
+        "file_data": file_data,
```
Comment (manycoding) [inline lib/anydocs/processAnyDocumentBuffer.js:24]: file buffer in general is binary, file_data is base64

File: lib/bankstatements/processBankStatement.js [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+ */
+Client.prototype.process_bank_statement = async function (
+    file_path,
+    bounding_boxes = false,
+    confidence_details = false,
+    {...kwargs} = {}
+) {
+
+    let file_name = path.basename(file_path);
+    const image_file = fs.readFileSync(file_path, {encoding: 'base64'});
```
Comment (manycoding) [inline lib/bankstatements/processBankStatement.js:24]: same, let's do multipart

File: lib/bankstatements/processBankStatementBuffer.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ * @param {String} file_name The file name including the extension
+ * @param {String} file_buffer Buffer of a file to submit for data extraction
+ * @param {boolean} bounding_boxes A field used to determine whether to return bounding_box and bounding_region for extracted fields in the Document response.
+ * @param {boolean} confidence_details A field used to determine whether to return the score and ocr_score fields in the Document response.
+ * @param {Object} kwargs Additional request parameters
+ * @returns {JSON} Data extracted from the document
+ */
+Client.prototype.process_bank_statement_from_buffer = async function (
+    file_name,
+    file_buffer,
```
Comment (manycoding) [inline lib/bankstatements/processBankStatementBuffer.js:15]: same

File: lib/businesscards/processBusinessCardBuffer.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @memberof Client
+ * @param {String} file_name The file name including the extension
+ * @param {String} file_buffer Buffer of a file to submit for data extraction
+ * @param {Object} kwargs Additional request parameters
+ * @returns {JSON} Data extracted from the document
+ */
+Client.prototype.process_business_card_from_buffer = async function (
+    file_name,
+    file_buffer,
```
Comment (manycoding) [inline lib/businesscards/processBusinessCardBuffer.js:13]: same

File: lib/client/checkMimeType.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
+const Client = require('../client/constructor');
+/**
+ * Checks if a Base64 string contains any of the predefined MIME types.
+ *
+ * @param {string} base64String - The Base64 encoded string with a MIME type prefix.
+ * @returns {boolean} - Returns true if any predefined MIME type is found, false otherwise.
+ */
+Client.prototype.check_mime_type = function (base64String) {
+    const mimeTypes = [
+        'image/png',
```
Comment (manycoding) [inline lib/client/checkMimeType.js:10]: I think there's more, like webp. You can just check if it starts with `data:` and has `;base64,`

File: lib/client/request.js (hunk 5 above)
Comment (manycoding) [reply lib/client/request.js:54]: so where are the details?
I don't understand what's the different between response?.response?.data?.error; and response?.message

File: lib/types/BillTo.ts (hunk 9 above)
Comment (manycoding) [reply lib/types/BillTo.ts:3]: ?

**Review by manycoding — CHANGES_REQUESTED** [review]
Two biggest points:
1. it's ok to allow clients to pass their base64, but we should not use base64 under the hood since it's a bit slower
2. file data is not a buffer

**Review by manycoding — APPROVED** [review]

**Pull request merged**